File: ClientRuntimes/NodeJS/ClientRuntime.Azure/lib/azureServiceClient.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 
'use strict';

var util = require('util');
var async = require('async');
var msrest = require('ms-rest');
var PollingState = require('./pollingState');
var LroStates = require('./constants').LongRunningOperationStates;
var WebResource = msrest.WebResource;

/**
 * @class
 * Initializes a new instance of the AzureServiceClient class.
 * @constructor
 * @param {object} [credentials] - BasicAuthenticationCredentials or 
 * TokenCredentials object used for authentication.  
 * 
 * @param {object} options - The parameter options used by ServiceClient
 * 
 * @param {Array} [options.longRunningOperationRetryTimeoutInSeconds] - Retry timeout
 * 
 */
function AzureServiceClient(credentials, options) {
  if (credentials === null || credentials === undefined ||
    credentials.subscriptionId === null || credentials.subscriptionId === undefined || 
    typeof credentials.subscriptionId !== 'string') {
    throw new Error('Azure clients require credentials with a valid subscriptionId');
  }
  
  AzureServiceClient['super_'].call(this, credentials, options);
  if (options) {
    this.longRunningOperationRetryTimeoutInSeconds = 
      options.longRunningOperationRetryTimeoutInSeconds;
  }
}

util.inherits(AzureServiceClient, msrest.ServiceClient);

/**
 * Poll Azure long running PUT operation.
 * @param {object} [resultOfInitialRequest] - Response of the initial request for the long running operation.
 * @param {function} [poller] - Poller function used to poll operation result.
 */
AzureServiceClient.prototype.getPutOperationResult = function (resultOfInitialRequest, poller, callback) {
  var self = this;
  if (!callback) {
    throw new Error('Missing callback');
  }
  
  if (!resultOfInitialRequest) {
    return callback(new Error('Missing resultOfInitialRequest parameter'));
  }
  
  if (!poller) {
    return callback(new Error('Missing poller parameter'));
  }
  
  if (resultOfInitialRequest.response.statusCode !== 200 &&
      resultOfInitialRequest.response.statusCode !== 201 &&
      resultOfInitialRequest.response.statusCode !== 202) {
    return callback(new Error(util.format('Unexpected polling status code from long running operation \'%s\'', 
      resultOfInitialRequest.response.statusCode)));
  }
  
  var pollingState = new PollingState(resultOfInitialRequest, this.longRunningOperationRetryTimeoutInSeconds);
  
  async.whilst(
    //while condition
    function () {
      var finished = [LroStates.Succeeded, LroStates.Failed, LroStates.Canceled].some(function (e) {
        return pollingState.status === e;
      });
      return !finished;
    },
    //while loop body
    function (callback) {
      if (pollingState.response.headers['azure-asyncoperation']) {
        self._updateStateFromAzureAsyncOperationHeader(pollingState, function (err) {
          if (err) return callback(err);
          setTimeout(callback, pollingState.getTimeout());
        });
      } else if (pollingState.response.headers['location']) {
        self._updateStateFromLocationHeaderOnPut(pollingState, function (err) {
          if (err) return callback(err);
          setTimeout(callback, pollingState.getTimeout());
        });
      } else {
        self._updateStateFromGetResourceOperation(poller, pollingState, function (err) {
          if (err) return callback(err);
          setTimeout(callback, pollingState.getTimeout());
        });
      }
    },
    //when done
    function (err) {
      if (pollingState.status === LroStates.Succeeded) {
        if (!pollingState.resource) {
          self._updateStateFromGetResourceOperation(poller, pollingState, function (err) {
            return callback(err, pollingState.getOperationResponse());
          });
        } else {
          return callback(null, pollingState.getOperationResponse()); 
        }
      } else {
        return callback(pollingState.getCloudError(err));
      }
    });
};

/**
 * Poll Azure long running POST or DELETE operations.
 * @param {object} [resultOfInitialRequest] - result of the initial request.
 */
AzureServiceClient.prototype.getPostOrDeleteOperationResult = function (resultOfInitialRequest, callback) {
  var self = this;
  if (!callback) {
    throw new Error('Missing callback');
  }
  
  if (!resultOfInitialRequest) {
    return callback(new Error('Missing resultOfInitialRequest parameter'));
  }
  
  if (!resultOfInitialRequest.response) {
    return callback(new Error('Missing resultOfInitialRequest.response'));
  }
  
  if (resultOfInitialRequest.response.statusCode !== 200 &&
      resultOfInitialRequest.response.statusCode !== 202 &&
      resultOfInitialRequest.response.statusCode !== 204) {
    return callback(new Error(util.format('Unexpected polling status code from long running operation \'%s\'', 
      resultOfInitialRequest.response.statusCode)));
  }
  
  var pollingState = new PollingState(resultOfInitialRequest, this.longRunningOperationRetryTimeoutInSeconds);
  
  async.whilst(
    function () {
      var finished = [LroStates.Succeeded, LroStates.Failed, LroStates.Canceled].some(function (e) {
        return e === pollingState.status;
      });
      return !finished;
    },
    function (callback) {
      if (pollingState.response.headers['azure-asyncoperation']) {
        self._updateStateFromAzureAsyncOperationHeader(pollingState, function (err) {
          if (err) return callback(err);
          setTimeout(callback, pollingState.getTimeout());
        });
      } else if (pollingState.response.headers['location']) {
        self._updateStateFromLocationHeaderOnPostOrDelete(pollingState, function (err) {
          if (err) return callback(err);
          setTimeout(callback, pollingState.getTimeout());
        });
      } else {
        return callback(new Error('Location header is missing from long running operation.'));
      }
    },
    function (err) {
      if (pollingState.status === LroStates.Succeeded ) {
        return callback(null, pollingState.getOperationResponse());
      } else {
        return callback(pollingState.getCloudError(err));
      }
    });
};

/**
 * Retrieve operation status by polling from 'azure-asyncoperation' header.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromAzureAsyncOperationHeader = function (pollingState, callback) {
  this._getStatus(pollingState.response.headers['azure-asyncoperation'], function (err, result) {
    if (err) return callback(err);
    
    if (!result.body || !result.body.status) {
      return callback(new Error('The response from long running operation does not contain a body.'));
    }
    
    pollingState.status = result.body.status;
    pollingState.error = result.body.error;
    pollingState.response = result.response;
    pollingState.request = result.request;
    pollingState.resource = null;//TODO: confirm we do want to clear.
    callback(null);
  });
};

/**
 * Retrieve PUT operation status by polling from 'location' header.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromLocationHeaderOnPut = function (pollingState, callback) {
  this._getStatus(pollingState.response.headers['location'], function (err, result) {
    if (err) return callback(err);
    
    pollingState.response = result.response;
    pollingState.request = result.request;
    
    var statusCode = result.response.statusCode;
    if (statusCode === 202) {
      pollingState.status = LroStates.InProgress;
    }
    else if (statusCode === 200 ||
             statusCode === 201) {
      
      if (!result.body) {
        return callback(new Error('The response from long running operation does not contain a body.'));
      }
      
      // In 202 pattern on PUT ProvisioningState may not be present in 
      // the response. In that case the assumption is the status is Succeeded.
      if (result.body.properties && result.body.properties.provisioningState) {
        pollingState.status = result.body.properties.provisioningState;
      }
      else {
        pollingState.status = LroStates.Succeeded;
      }
      
      pollingState.error = {
        code: pollingState.Status,
        message: util.format('Long running operation failed with status \'%s\'.', pollingState.status)
      };
      pollingState.resource = result.body;
    }
    callback(null);
  });
};

/**
 * Retrieve POST or DELETE operation status by polling from 'location' header.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromLocationHeaderOnPostOrDelete = function (pollingState, callback) {
  this._getStatus(pollingState.response.headers['location'], function (err, result) {
    if (err) return callback(err);
    
    pollingState.response = result.response;
    pollingState.request = result.request;
    
    var statusCode = result.response.statusCode;
    if (statusCode === 202) {
      pollingState.status = LroStates.InProgress;
    }
    else if (statusCode === 200 ||
             statusCode === 201 ||
             statusCode === 204) {
      pollingState.status = LroStates.Succeeded;
      pollingState.resource = result.body;
    }
    callback(null);
  });
};

/**
 * Retrieve operation status by invoking the poller function.
 * @param {function} [poller] - Poller function used to poll operation result.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromGetResourceOperation = function (poller, pollingState, callback) {
  var pollerCallback = function (err, result) {
    if (err) return callback(err);
    if (!result.body) {
      return callback(new Error('The response from long running operation does not contain a body.'));
    }
    
    if (result.body.properties.provisioningState) {
      pollingState.status = result.body.properties.provisioningState;
    } else {
      pollingState.status = LroStates.Succeeded;
    }
    
    //we might not throw an error, but initialize here just in case.
    pollingState.error = {
      code: pollingState.status,
      message: util.format('Long running operation failed with status \'%s\'.', pollingState.status)
    };
    
    pollingState.response = result.response;
    pollingState.request = result.request;
    pollingState.resource = result.body;
    
    //nothing to return, the 'pollingState' has all the info we care.
    callback(null);
  };
  poller.apply(this, [pollerCallback]);
};

/**
 * Retrieve operation status by querying the operation URL.
 * @param {string} [operationUrl] - URL used to poll operation result.
 */
AzureServiceClient.prototype._getStatus = function (operationUrl, callback) {
  var self = this;
  if (!operationUrl) {
    return callback(new Error('operationUrl cannot be null.'));
  }
  
  // Construct URL
  var requestUrl = operationUrl.replace(' ', '%20');
  
  // Create HTTP transport objects
  var httpRequest = new WebResource();
  httpRequest.method = 'GET';
  httpRequest.headers = {};
  httpRequest.url = requestUrl;

  // Send Request
  return self.pipeline(httpRequest, function (err, response, responseBody) {
    if (err) {
      return callback(err);
    }
    var statusCode = response.statusCode;
    if (statusCode !== 200 && statusCode !== 201 && statusCode !== 202 && statusCode !== 204) {
      var error = new Error(util.format('Invalid status code with response body "%s" occurred ' + 
        'when polling for operation status.', responseBody));
      error.statusCode = response.statusCode;
      error.request = httpRequest;
      error.response = response;
      if (responseBody === '') responseBody = null;
      try {
        error.body = JSON.parse(responseBody);

      } catch (badResponse) {
        error.message += util.format(' Could not deserialize error response body - "%s" -.', responseBody);
        error.body = responseBody;
      }

      return callback(error);
    }
    // Create Result
    var result = new msrest.HttpOperationResponse();
    result.request = httpRequest;
    result.response = response;
    if (responseBody === '') responseBody = null;
    try {
      result.body = JSON.parse(responseBody);
    } catch (deserializationError) {
      var parseError = new Error(util.format('Error "%s" occurred in deserializing the response body - "%s" -' + 
        ' when polling for operation status.', deserializationError, responseBody));
      parseError.request = httpRequest;
      parseError.response = response;
      parseError.body = responseBody;
      return callback(parseError);
    }

    return callback(null, result);
  });
};

module.exports = AzureServiceClient;

File: ClientRuntimes/Ruby/ms-rest-azure/lib/client_runtime_azure/azure_service_client.rb
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

module ClientRuntimeAzure
  #
  # Class which represents a point of access to the REST API.
  #
  class AzureServiceClient < ClientRuntime::ServiceClient

    # @return [String] api version of the Azure in string format.
    attr_accessor :api_version

    #
    # Creates and initialize new instance of the ServiceClient class.
    #
    def initialize(credentials, options)
      super(credentials, options)
      @api_version = '2015-05-01-preview'
    end

    #
    # Retrieves the result of 'PUT' operation. Perfroms polling of required.
    # @param azure_response [ClientRuntimeAzure::AzureOperationResponse] response from Azure service.
    # @param get_operation_block [Proc] custom method for polling.
    # @param custom_headers [Hash] custom HTTP headers to apply to HTTP requests.
    #
    # @return [Concurrent::Promise] promise to return response from Azure service.
    def get_put_operation_result(azure_response, get_operation_block, custom_headers, custom_deserialization_block)
      fail CloudError if azure_response.nil?
      fail CloudError if get_operation_block.nil?

      if (azure_response.response.code != "200" &&
          azure_response.response.code != "201" &&
          azure_response.response.code != "202")
        fail CloudError
      end

      polling_state = PollingState.new(azure_response, 1) # TODO: add timeout

      if (!AsyncOperationStatus.is_terminal_status(polling_state.status))
        task = Concurrent::TimerTask.new do
          begin
            if !polling_state.azure_async_operation_header_link.nil?
              update_state_from_azure_async_operation_header(polling_state, custom_headers)
            elsif !polling_state.location_header_link.nil?
              update_state_from_location_header_on_put(polling_state, custom_headers, custom_deserialization_block)
            else
              update_state_from_get_resource_operation(get_operation_block, polling_state)
            end

            if (AsyncOperationStatus.is_terminal_status(polling_state.status))
              task.shutdown
            end
          rescue Exception => e
            task.shutdown
            e
          end
        end

        task.execution_interval = polling_state.get_delay_in_milliseconds()
        task.execute
        task.wait_for_termination

        polling_error = task.value
        fail polling_error if polling_error.is_a?(Exception)
      end

      promise = Concurrent::Promise.new do
        if (AsyncOperationStatus.is_successful_status(polling_state.status) && polling_state.resource.nil?)
          update_state_from_get_resource_operation(get_operation_block, polling_state)
        end

        if (AsyncOperationStatus.is_failed_status(polling_state.status))
          fail CloudError # TODO: proper error
        end

        polling_state.get_operation_response
      end

      promise.execute
    end

    #
    # Retrieves the result of 'POST' or 'DELETE' operations. Perfroms polling of required.
    # @param azure_response [ClientRuntimeAzure::AzureOperationResponse] response from Azure service.
    # @param custom_headers [Proc] custom method for polling.
    # @param custom_deserialization_block [Proc] custom logic for response deserialization.
    #
    # @return [Concurrent::Promise] promise to return response from Azure service.
    def get_post_or_delete_operation_result(azure_response, custom_headers, custom_deserialization_block)
      fail CloudError if azure_response.nil?
      fail CloudError if azure_response.response.nil?

      if (azure_response.response.code != "200" &&
          azure_response.response.code != "202" &&
          azure_response.response.code != "204")
        fail CloudError
      end

      polling_state = PollingState.new(azure_response, 1) # TODO: add timeout

      if (!AsyncOperationStatus.is_terminal_status(polling_state.status))
        task = Concurrent::TimerTask.new do
          begin
            if !polling_state.azure_async_operation_header_link.nil?
              update_state_from_azure_async_operation_header(polling_state, custom_headers)
            elsif !polling_state.location_header_link.nil?
              update_state_from_location_header_on_post_or_delete(polling_state, custom_headers, custom_deserialization_block)
            else
              task.shutdown
              fail CloudError
            end

            if (AsyncOperationStatus.is_terminal_status(polling_state.status))
              task.shutdown
            end
          rescue Exception => e
            task.shutdown
            e
          end
        end

        task.execution_interval = polling_state.get_delay_in_milliseconds()
        task.execute
        task.wait_for_termination

        polling_error = task.value
        fail polling_error if polling_error.is_a?(Exception)
      end

      promise = Concurrent::Promise.new do
        if (AsyncOperationStatus.is_failed_status(polling_state.status))
          fail CloudError # TODO: proper error
        end

        polling_state.get_operation_response
      end

      promise.execute
    end

    #
    # Updates polling state based on location header for PUT HTTP requests.
    # @param get_operation_block [Proc] custom method for polling.
    # @param polling_state [ClientRuntimeAzure::PollingState] polling state to update.
    def update_state_from_get_resource_operation(get_operation_block, polling_state)
      result = get_operation_block.call().value!

      fail CloudError if (result.body.nil?)

      if (result.body.respond_to?(:properties) && result.body.properties.respond_to?(:provisioning_state) && !result.body.properties.provisioning_state.nil?)
        polling_state.status = result.body.properties.provisioning_state
      else
        polling_state.status = AsyncOperationStatus::SUCCESS_STATUS
      end

      # TODO: fill the polling_state.error

      polling_state.update_response(result.response)
      polling_state.request = result.request
      polling_state.resource = result.body
    end

    #
    # Updates polling state based on location header for PUT HTTP requests.
    # @param polling_state [ClientRuntimeAzure::PollingState] polling state to update.
    # @param custom_headers [Hash] custom headers to apply to HTTP request.
    # @param custom_deserialization_block [Proc] custom deserialization method for parsing response.
    def update_state_from_location_header_on_put(polling_state, custom_headers, custom_deserialization_block)
      result = get_async_common(polling_state.location_header_link, custom_headers, custom_deserialization_block).value!

      polling_state.update_response(result.response)
      polling_state.request = result.request

      status_code = result.response.code

      if (status_code == "202")
        polling_state.status = AsyncOperationStatus::IN_PROGRESS_STATUS
      elsif (status_code == "200" || status_code == "201")
        fail CloudError if (result.body.nil?)
          # TODO elaborate error

      if (result.body.respond_to?(:properties) && result.body.properties.respond_to?(:provisioning_state) && !result.body.properties.provisioning_state.nil?)
          polling_state.status = result.body.properties.provisioning_state
        else
          polling_state.status = AsyncOperationStatus::SUCCESS_STATUS
        end

        # TODO add filling of the polling_state.error

        polling_state.resource = result.body
      end
    end

    #
    # Updates polling state from Azure async operation header.
    # @param polling_state [ClientRuntimeAzure::PollingState] polling state.
    # @param custom_headers [Hash] custom headers to apply to HTTP request.
    def update_state_from_azure_async_operation_header(polling_state, custom_headers)
      result = get_async(polling_state.azure_async_operation_header_link, custom_headers).value!

      # TODO elaborate error
      fail CloudError if (result.body.nil? || result.body.status.nil?)

      polling_state.status = result.body.status
      # TODO: fill the error
      # polling_state.error = nil
      polling_state.response = result.response
      polling_state.request = result.request
      polling_state.resource = nil
    end

    #
    # Updates polling state based on location header for POST and DELETE HTTP requests.
    # @param polling_state [ClientRuntimeAzure::PollingState] [description]
    # @param custom_headers [Hash] custom headers to apply to HTTP requests.
    # @param custom_deserialization_block [Proc] custom deserialization method for parsing response.
    def update_state_from_location_header_on_post_or_delete(polling_state, custom_headers, custom_deserialization_block)
      result = get_async_common(polling_state.location_header_link, custom_headers, custom_deserialization_block).value!

      polling_state.update_response(result.response)
      polling_state.request = result.request

      status_code = result.response.code

      if (status_code == "202")
        polling_state.status = AsyncOperationStatus::IN_PROGRESS_STATUS
      elsif (status_code == "200" || status_code == "201" || status_code == "204")
        polling_state.status = AsyncOperationStatus::SUCCESS_STATUS
        polling_state.resource = result.body
      end
    end

    #
    # Retrives data by given URL.
    # @param operation_url [String] the URL.
    # @param custom_headers [String] headers to apply to the HTTP request.
    #
    # @return [ClientRuntimeAzure::AsyncOperationStatus] Deserialized response wrapped by AsyncOperationStatus.
    def get_async(operation_url, custom_headers)
      fail CloudError if operation_url.nil?

      # TODO: add url encoding.
      url = URI(operation_url)

      fail URI::Error unless url.to_s =~ /\A#{URI::regexp}\z/

      # Create HTTP transport objects
      http_request = Net::HTTP::Get.new(url.request_uri)
      http_request.add_field('Content-Type', 'application/json')

      # TODO: add custom headers

      # Send Request
      promise = Concurrent::Promise.new { self.make_http_request(http_request, url) }

      promise = promise.then do |http_response|
        status_code = http_response.code.to_i
        response_content = http_response.body

        if (status_code != 200 && status_code != 201 && status_code != 202 && status_code != 204)
          fail CloudError
        end

        result = ClientRuntime::HttpOperationResponse.new(http_request, http_response, http_response.body)

        begin
          parsed_response = JSON.load(response_content) unless response_content.to_s.empty?
          result.body = AsyncOperationStatus.deserialize_object(parsed_response)
        rescue Exception => e
          fail ClientRuntime::DeserializationError.new("Error occured in deserializing the response", e.message, e.backtrace, response_content)
        end

        result
      end

      promise.execute
    end

    #
    # Retrives data by given URL.
    # @param operation_url [String] the URL.
    # @param custom_headers [String] headers to apply to the HTTP request.
    # @param custom_deserialization_block [Proc] function to perform deserialization of the HTTP response.
    #
    # @return deserialized response (return type may differ depending on response)
    def get_async_common(operation_url, custom_headers, custom_deserialization_block)
      fail CloudError if operation_url.nil?

      # TODO: add url encoding.
      url = URI(operation_url)

      fail URI::Error unless url.to_s =~ /\A#{URI::regexp}\z/

      # Create HTTP transport objects
      http_request = Net::HTTP::Get.new(url.request_uri)
      http_request.add_field('Content-Type', 'application/json')

      # TODO: add custom headers

      # Send Request
      promise = Concurrent::Promise.new { self.make_http_request(http_request, url) }

      promise = promise.then do |http_response|
        status_code = http_response.code.to_i
        response_content = http_response.body

        if (status_code != 200 && status_code != 201 && status_code != 202 && status_code != 204)
          fail CloudError
        end

        result = ClientRuntime::HttpOperationResponse.new(http_request, http_response, http_response.body)

        begin
          parsed_response = JSON.load(response_content) unless response_content.to_s.empty?
          if (parsed_response && !custom_deserialization_block.nil?)
            parsed_response = custom_deserialization_block.call(parsed_response)
          end
        rescue Exception => e
          fail ClientRuntime::DeserializationError.new("Error occured in deserializing the response", e.message, e.backtrace, response_content)
        end

        result.body = parsed_response

        result
      end

      promise.execute
    end
  end

end
